File: baekjoon/dfs/11724/11724.java
import java.util.*;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.BufferedWriter;
import java.io.OutputStreamWriter;

public class Main{
  static int N;
  static int[][] graph;
  static int[] visited;

  static void dfs(int v){
    visited[v] = 1;
    for(int i=1;i<=N;i++){
      if(graph[v][i] == 1 && visited[i] == 0)
        dfs(i);
    }
  }

  public static void main(String []args) throws Exception{
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    String[] tmp = br.readLine().split(" ");
    N = Integer.parseInt(tmp[0]);
    int M = Integer.parseInt(tmp[1]);
    int cnt = 0;

    graph = new int[N+1][N+1];
    visited = new int[N+1];

    for(int i=0;i<M;i++){
      tmp = br.readLine().split(" ");
      int row = Integer.parseInt(tmp[0]), col = Integer.parseInt(tmp[1]);
      graph[row][col] = 1; graph[col][row] = 1;
    }

    for(int i=1;i<=N;i++){
      if(visited[i] == 0){
        ++cnt;
        dfs(i);
      }
    }

    bw.write(cnt+"\n");
    bw.flush();
  } 
}

File: baekjoon/dfs/11724/11724.py
import sys

def dfs(v):
  global visited
  visited[v] = 1
  for i in range(1,N+1):
    if graph[v][i] == 1 and visited[i] == 0:
      dfs(i)

sys.setrecursionlimit(1100)
N, M = map(int,sys.stdin.readline().split())
graph = [[0]*(N+1) for _ in range(N+1)]
visited = [0]*(N+1)
cnt = 0

for _ in range(M):
  row, col = map(int, sys.stdin.readline().split())
  graph[row][col] = 1
  graph[col][row] = 1

for i in range(1,N+1):
  if visited[i] == 0:
    cnt += 1
    dfs(i)

print(cnt)
